eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelTypes.scala

```diff
@@ -298,6 +306,30 @@ final case class DATA_CLOSING(commitments: Commitments,
                               revokedCommitPublished: List[RevokedCommitPublished] = Nil) extends Data with HasCommitments {
   val spendingTxes = mutualClosePublished ::: localCommitPublished.map(_.commitTx).toList ::: remoteCommitPublished.map(_.commitTx).toList ::: nextRemoteCommitPublished.map(_.commitTx).toList ::: futureRemoteCommitPublished.map(_.commitTx).toList ::: revokedCommitPublished.map(_.commitTx)
   require(spendingTxes.nonEmpty, "there must be at least one tx published in this state")
+
+  override def htlcsToFailInOrigin(log: LoggingAdapter): Set[Long] = {
```

Comment:
This has nothing to do inside `DATA_CLOSING` :thinking:

Reply:
It's related to `DATA_CLOSING` in a sense that here: https://github.com/ACINQ/eclair/blob/b0716aea8839bb0593293e19fb6397a578365989/eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/PostRestartHtlcCleaner.scala#L325 `checkBrokenHtlcs` matches against data and applies this code if `DATA_CLOSING` is found.

This change basically removes a need for a match and opens a possibility for every `Data` type to provide its own `htlcsToFailInOrigin`. Currently it's only overridden in `DATA_CLOSING` but it would be convenient to have it in plugin `Data` types, too. This is where `HasAbstractCommitments` will come handy, plugin data types are supposed to extend it.

Reply:
Oh right, I reviewed this one too quickly. Will revisit.